File: api/gif.py
"""이미지 → GIF 변환기 API.

이미지 처리는 CPU를 많이 쓰므로 동시 작업 수를 제한한다.
def 엔드포인트라 FastAPI가 스레드풀에서 실행한다.
"""

import json
from threading import BoundedSemaphore
from typing import Literal, Optional
from urllib.parse import quote

from fastapi import APIRouter, File, Form, HTTPException, UploadFile
from fastapi.responses import Response

from services import gif_maker as gm

router = APIRouter(prefix="/api/gif", tags=["gif"])
_slots = BoundedSemaphore(2)
HEX = r"^#[0-9A-Fa-f]{6}$"


def _read(upload: Optional[UploadFile]) -> Optional[bytes]:
    if upload is None:
        return None
    data = upload.file.read(gm.MAX_UPLOAD_BYTES + 1)
    if len(data) > gm.MAX_UPLOAD_BYTES:
        raise HTTPException(413, "이미지는 15MB 이하만 올릴 수 있습니다.")
    return data or None


def _run(fn, *args):
    if not _slots.acquire(timeout=60):
        raise HTTPException(503, "지금 변환 작업이 많아요. 잠시 후 다시 시도해 주세요.")
    try:
        return fn(*args)
    except ValueError as e:
        raise HTTPException(400, str(e))
    finally:
        _slots.release()


@router.post("/preview")
def preview(
    image: UploadFile = File(...),
    background: Optional[UploadFile] = File(None),
    remove_bg: bool = Form(False),
    use_background: bool = Form(False),
    canvas_color: str = Form("#F3F5F8", pattern=HEX),
):
    png = _run(gm.preview, _read(image), _read(background), remove_bg, use_background, canvas_color)
    return Response(png, media_type="image/png", headers={"Cache-Control": "no-store"})


@router.post("/generate")
def generate(
    image: UploadFile = File(...),
    background: Optional[UploadFile] = File(None),
    remove_bg: bool = Form(False),
    use_background: bool = Form(False),
    canvas_color: str = Form("#F3F5F8", pattern=HEX),
    use_drive: bool = Form(True),
    use_zoom: bool = Form(False),
    use_bounce: bool = Form(True),
    use_shake: bool = Form(False),
    use_tilt: bool = Form(False),
    use_pan: bool = Form(False),
    direction: Literal["왼쪽 → 오른쪽", "오른쪽 → 왼쪽"] = Form("왼쪽 → 오른쪽"),
    travel_strength: int = Form(20, ge=5, le=35),
    perspective_strength: int = Form(18, ge=0, le=30),
    speed: int = Form(3, ge=1, le=5),
    # 폼 값은 문자열이라 int Literal로는 검증되지 않으므로 int로 받고 아래에서 허용값을 확인
    frame_count: int = Form(20),
    max_width: int = Form(800),
    loop_style: Literal["일반 반복", "부메랑"] = Form("일반 반복"),
    brightness: int = Form(100, ge=80, le=120),
    contrast: int = Form(100, ge=80, le=120),
):
    if frame_count not in gm.FRAME_OPTIONS or max_width not in gm.WIDTH_OPTIONS:
        raise HTTPException(400, "프레임 수 또는 출력 폭 값이 올바르지 않습니다.")
    opts = gm.GifOptions(
        remove_bg=remove_bg, use_background=use_background, canvas_color=canvas_color,
        use_drive=use_drive, use_zoom=use_zoom, use_bounce=use_bounce, use_shake=use_shake, use_tilt=use_tilt, use_pan=use_pan,
        direction=direction, travel_strength=travel_strength, perspective_strength=perspective_strength,
        speed=speed, frame_count=frame_count, max_width=max_width, loop_style=loop_style,
        brightness=brightness, contrast=contrast,
    )
    data, meta = _run(gm.make_gif, _read(image), _read(background), opts)
    # 헤더는 latin-1만 허용되므로 JSON을 URL 인코딩해서 보낸다
    return Response(data, media_type="image/gif",
                    headers={"X-Gif-Meta": quote(json.dumps(meta, ensure_ascii=False)), "Cache-Control": "no-store"})

File: web/src/lib/gif.ts
// FastAPI /api/gif 클라이언트 (multipart 업로드 → 이미지 바이너리 응답)
import { apiFetch } from "@/lib/access";

export const API_URL = (process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000").replace(/\/$/, "");
export const MAX_UPLOAD_MB = 15;

export type GifOptions = {
  remove_bg: boolean;
  use_background: boolean;
  canvas_color: string;
  use_drive: boolean;
  use_zoom: boolean;
  use_bounce: boolean;
  use_shake: boolean;
  use_tilt: boolean;
  use_pan: boolean;
  direction: "왼쪽 → 오른쪽" | "오른쪽 → 왼쪽";
  travel_strength: number;
  perspective_strength: number;
  speed: number;
  frame_count: number;
  max_width: number;
  loop_style: "일반 반복" | "부메랑";
  brightness: number;
  contrast: number;
};

export type GifMeta = { frames: number; width: number; height: number; duration: number; loop_style: string; bytes: number };

export const DEFAULT_OPTIONS: GifOptions = {
  remove_bg: false,
  use_background: false,
  canvas_color: "#F3F5F8",
  use_drive: true,
  use_zoom: false,
  use_bounce: true,
  use_shake: false,
  use_tilt: false,
  use_pan: false,
  direction: "왼쪽 → 오른쪽",
  travel_strength: 20,
  perspective_strength: 18,
  speed: 3,
  frame_count: 20,
  max_width: 800,
  loop_style: "일반 반복",
  brightness: 100,
  contrast: 100,
};

function formData(image: File, background: File | null, fields: Record<string, string | number | boolean>) {
  const fd = new FormData();
  fd.append("image", image);
  if (background) fd.append("background", background);
  for (const [k, v] of Object.entries(fields)) fd.append(k, String(v));
  return fd;
}

async function post(path: string, fd: FormData, signal?: AbortSignal) {
  const res = await apiFetch(`${API_URL}/api/gif${path}`, { method: "POST", body: fd, signal });
  if (!res.ok) {
    let message = `서버 응답 오류 (${res.status})`;
    try {
      const j = await res.json();
      if (typeof j?.detail === "string") message = j.detail;
      else if (Array.isArray(j?.detail)) message = "입력값을 확인해 주세요.";
    } catch {}
    throw new Error(message);
  }
  return res;
}

export const gifApi = {
  async preview(image: File, background: File | null, o: Pick<GifOptions, "remove_bg" | "use_background" | "canvas_color">, signal?: AbortSignal) {
    const res = await post("/preview", formData(image, background, o), signal);
    return res.blob();
  },
  async generate(image: File, background: File | null, o: GifOptions, signal?: AbortSignal) {
    const res = await post("/generate", formData(image, background, o), signal);
    const raw = res.headers.get("X-Gif-Meta");
    const meta = raw ? (JSON.parse(decodeURIComponent(raw)) as GifMeta) : null;
    return { blob: await res.blob(), meta };
  },
};
